sys/kern/exception.c

```diff
@@ -17,18 +21,50 @@ void on_exc_leave(void) {
   }
 }
 
-void on_user_exc_leave(void) {
-  thread_t *td = thread_self();
+static void set_syscall_retval(mcontext_t *ctx, syscall_result_t *result,
+                               int sig) {
+  int error = result->error;
+  proc_t *p = proc_self();
+
+  if (error == EJUSTRETURN)
+    return;
 
+  if (error == ERESTARTSYS || error == ERESTARTNOHAND) {
+    if (!sig || (error == ERESTARTSYS &&
+                 (p->p_sigactions[sig].sa_flags & SA_RESTART))) {
+      mcontext_restart_syscall(ctx);
+      return;
+    }
+    error = EINTR;
+  }
```

Comment:
It would be nice if it could be simplified. Otherwise, it's worth adding a comment here.

Reply:
I think I've made it more readable now. Do you agree?

Reply:
Yes, thanks.
You may consider adding `{`, `}` for maximum readability.

Reply:
Done.